misc-scripts/gifts_import/populate_protein_feature.pl

```diff
@@ -0,0 +1,350 @@
+    eval{
+    my $affected = $protfeature_dbh->do($insert_stmt);
+    $rows_inserted += $affected;
+    };
+    if($@){
+      print "$@\n";
+      push(@failed_stmts, $@);
+    }
+    
+  }
+  print Dumper(\@failed_stmts);
+  print "INSERT END\n";
+  }
+
+  print "Number of rows inserted into protein_feature $protfeatureDB->{'dbname'}.protein_feature ", $rows_inserted, "\n";
+  return $rows_inserted;
+  
+  $gifts_dbh->disconnect();
+  $protfeature_dbh->disconnect();
+  
+}
+
+sub annotate_ref{
+  my $ref = shift;
+  my $ensp_id = $ref->[0];
+  my $ensp = $translation_adaptor->fetch_by_stable_id( $ensp_id ) ;
+  if($ensp){
+  $ref->[0] = $ensp->dbID;
+  $ref->[1] = 1;
+  $ref->[2] = $ensp->length()-1;
+  
+  }
+  
+  return $ref;
+}
+
+sub build_insert_sql {
+  my ($gifts_dbh, $protfeature_dbh) = @_;
+  
+  #get analysis id
+  my $analysis_id = get_analysis_id("gifts_import");
+  print("Analysis id $analysis_id \n");
+  
+  my $select_sql = qq{SELECT DISTINCT ensp_id, 0, 0, uniprot_acc, $analysis_id as analysis_id, mdz, 'mdtag' FROM ensp_u_cigar};
+  $select_sql = $limit ? $select_sql . " limit 10" : $select_sql;
+  
+  my $insert_sql = qq{INSERT INTO $protfeatureDB->{'dbname'}.protein_feature(translation_id, seq_start, seq_end, hit_name, analysis_id, cigar_line, align_type) VALUES };
+
+
+  my $import_sth = $gifts_dbh->prepare($select_sql);
+  $import_sth->execute;
+  my $rows = $import_sth->rows;
+
+  my $start = 0;
+  my $insert_values;
+  my $max_rows_in_insert = 9999;    #Batch size of 10,000
+  # my $max_rows_in_insert = 10;    #Batch size of 10
+  my $rows_in_insert     = 0;
+
+  my @insert_container = ();
+  my $miss = 1;
+  while ( my $ref = $import_sth->fetchrow_arrayref ) {
+     my $not_ensp =1;
+     if($ref->[0] !~ /^ENSP.*/){
+      warn "\t\t$not_ensp ensp id not in right format for ", $ref->[0], " Uniprot id ", $ref->[3], "\n";
+      $not_ensp++;
+      next;
+    }
+    $ref = annotate_ref($ref);
+    if($ref->[0] =~ /^ENSP.*/){
+      warn "$miss No translation record found in core translation for ", $ref->[0], " Uniprot id ", $ref->[3], "\n";
+      $miss++;
+      next;
+    }
+    
+    if ( $rows_in_insert < $max_rows_in_insert ) {
+      $insert_values .= ',' if $start++;
+      $insert_values .= '(' . ( join( ",", map { $gifts_dbh->quote($_) } @{$ref} ) ) . ')';
+      $rows_in_insert++;
+    }
+    else {
+      $insert_values .= ',' if $start++;
+      $insert_values .= '(' . ( join( ",", map { $gifts_dbh->quote($_) } @{$ref} ) ) . ')';
+      my $insert_sql_with_values = $insert_sql . $insert_values;
+      push( @insert_container, $insert_sql_with_values );
+
+      $insert_values  = undef;
+      $rows_in_insert = 0;
+      $start          = 0;
+  }
+}
+
+  if ($insert_values) {
+    my $insert_sql_with_values = $insert_sql . $insert_values;
+    push( @insert_container, $insert_sql_with_values );
+  }
+  
+  print "Number of rows fetched from GIFTS $rows \n";
+  return \@insert_container;
```

Comment:
How large is this array going to get over the next few years? I don't see any version in the SELECT that originally populates it. Is there a reason it's structured this way rather than directly selecting then inserting as you iterate over the result set?

In addition, the join(','...) with quote() isn't the safest method. If you are going to return an array of elements to insert, it'd probably be better to return a prepared insert statement with parametarized placeholder for the values plus an array of arrays of the values. Then use DBI to do all the proper quoting and filling in the placeholders in the execute() call by passing this array of values at that time.